code/espurna/tuya_util.h

```diff
@@ -12,114 +12,186 @@ Copyright (C) 2019 by Maxim Prokhorov <prokhorov dot max at outlook dot com>
 #include <algorithm>
 #include <vector>
 
-namespace Tuya {
+namespace tuya {
 
-    template <typename T>
-    class States {
+namespace util {
 
-        public:
+template <typename T>
+bool command_equals(const T& frame, Command command) {
+    return (frame.command() == static_cast<uint8_t>(command));
+}
 
-            struct Container {
-                uint8_t dp;
-                T value;
-            };
+}
 
-            using iterator = typename std::vector<Container>::iterator;
-            using const_iterator = typename std::vector<Container>::const_iterator;
+struct StateId {
+    StateId() = default;
+
+    void filter(bool value) {
+        _filter = value;
+    }
+
+    bool filter() {
+        return _filter;
+    }
+
+    uint8_t id() {
+        return _id;
+    }
+
+    StateId& operator=(uint8_t value) {
+        _id = value;
+        return *this;
+    }
+
+    explicit operator bool() {
+        return _id != 0u;
+    }
+
+private:
+    uint8_t _id { 0 };
+    bool _filter { false };
```

Comment:
For cold boot issue specifically
```suggestion
    bool _filter { true };
```
I wonder if this will work. This way it will ignore incoming value up until both relay and channel statuses are set

Reply:
Yep, that fixed it! After cold boot the light comes ON for about a second, then returns to OFF. `debug.buffer` gives:
```
[042390] [WEBSOCKET] Requested action: dbgcmd
[042395] [DEBUG] Buffer size: 4084 / 4096 bytes
[000127] [LIGHT] LIGHT_PROVIDER = 3
[000128] [LIGHT] Number of channels: 0
[000150] [RELAY] Number of relays: 1, boot mask: 0b1
[000150] [RELAY] #0 set to OFF
[000177] [BUTTON] Number of buttons: 1
[000183] [LED] Number of leds: 1
[000336] [RELAY] Number of relays: 1, boot mask: 0b1
[000337] [LIGHT] Number of channels: 1
[000347] [WIFI] Scanning
[000347] [RELAY] #0 set to OFF
[000957] [TUYA] <=: 55aa0007000501010001010f
[000969] [TUYA] <=: 55aa0007000802020004000000ff15
[001348] [RELAY] Relay mask: 0b0
[001983] [TUYA] =>: 55aa0006000501010001000d
[002015] [TUYA] <=: 55aa0007000501010001000e
[002026] [TUYA] <=: 55aa0007000501010001000e
[002038] [TUYA] <=: 55aa0007000802020004000000ff15
[002059] [TUYA] <=: 55aa0007000501010001000e
[002071] [TUYA] <=: 55aa00070008020200040000001026
[003502] [TUYA] =>: 55aa00000000ff
[042527] [TUYA] =>: 55aa00000000ff
[042552] [TUYA] <=: 55aa000000010101
```

Reply:
Hm. MCU turns the lights ON after ~1s all by itself though...
Just to understand wth it's doing, what happens after disabling transitions and commenting out `// updateSwitch(proto);` at line 264? Meaning, at around ~400ms we should send OFF, but never parse the incoming ON.

BTW, the 1s wait comes from the lights engine doing the whole 'steps' routine without ever checking that the channel value changes between each step, since we only ever get the final `=>: 55aa0006000501010001000d` after transition timer expires and does the 10th step. In earlier examples it waits exactly 500ms (`ltStep` default), right now it's 100ms. Will need to fix that

Reply:
Neither of the changes turn the light back OFF after a cold boot if transitions are turned OFF. The values are currently set (for no real reason) to 1600/200ms which works quite well when going from 0-100% and back.

The two videos show the LED state when re-applying mains power.

https://user-images.githubusercontent.com/3661426/104120516-6c459580-532f-11eb-839e-ba8ae1b3757c.mp4

https://user-images.githubusercontent.com/3661426/104120518-6d76c280-532f-11eb-8b65-013aea786913.mp4

Reply:
Why 1 second though :/ https://github.com/xoseperez/espurna/pull/2414/commits/24870953b6dfead1e9fd26265213c06521c9f18e adds early heartbeat message, maybe that's the case.

Transitions are also reworked to avoid pointlessly waiting for very small or zero steps.

Reply:
With all versions of espurna, a cold boot would bring the light ON. The only way to sync the ON/OFF state before was with the physical GPIO button and I think using the webui. MQTT wouldn't sync the switch/relay status.

Possibly completely not relevant..., a normal device would read the button high/low state at boot. As a button is not defined but I guess discovered, could the device be trying to read an unknown button state from the electronic noise at boot, I guess high at boot unless held low, and triggering the ON state?

Reply:
As I understood it - "The Wi-Fi module resets the network configuration by detecting the GPIO input pin." - MCU won't touch it by itself, but only sends the factory-selected configuration for the WiFi part
https://developer.tuya.com/en/docs/iot/device-development/embedded-software-development/mcu-development-access/wifi-mcu-sdk-solution/tuya-cloud-universal-serial-port-access-protocol?id=K9hhi0xxtn9cb#title-7-Querying%20working%20mode

Both GPIOs are already set via `btnGPIO0` and `ledGPIO0` in the logs above, but idk if that is really affects this in any way. You could `del` them both though just to try it out and set them later manually (since tuya.save won't have those, as discovery is never called with manual config)